sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -74,16 +80,28 @@ case class Mode(
     if (buffer.isEmpty) {
       return null
     }
-
+    val collationAwareBuffer =
+      if (!CollationFactory.fetchCollation(collationId).supportsBinaryEquality) {
+        val modeMap = buffer.toSeq.groupMapReduce {
+          case (key: String, _) =>
+            CollationFactory.getCollationKey(UTF8String.fromString(key), collationId)
+          case (key: UTF8String, _) =>
+            CollationFactory.getCollationKey(key, collationId)
+          case (key, _) => key
+        }(x => x)((x, y) => (x._1, x._2 + y._2)).values
+        modeMap
+      } else {
+        buffer
+      }
```

Comment:
```suggestion
    val collationAwareBuffer = child.dataType match {
      case st: StringType =>
          val modeMap = buffer.toSeq.groupMapReduce {
          case (key: String, _) =>
            CollationFactory.getCollationKey(UTF8String.fromString(key), collationId)
          case (key: UTF8String, _) =>
            CollationFactory.getCollationKey(key, collationId)
          case (key, _) => key
          }(x => x)((x, y) => (x._1, x._2 + y._2)).values
          modeMap
      case _ => buffer
    }
```

(something like this)

Reply:
@uros-db done